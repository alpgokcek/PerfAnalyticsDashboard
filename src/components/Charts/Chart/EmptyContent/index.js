import React from "react";

import "./empty-content.scss";

export default function EmptyContent() {
  return (
    <div className="empty-content">
      <div className="empty-content__wrapper">
        <img
          className="empty-content__wrapper-icon"
          src="assets/img/no-data.png"
          alt="no-data"
        />
        <div>Not content found.</div>
      </div>
    </div>
  );
}
