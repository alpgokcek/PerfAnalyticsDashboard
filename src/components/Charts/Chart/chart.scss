@import "../../../styles/variables.scss";
@import "../../../styles/placeholders.scss";

.chart {
  border: 1px solid $colorBlackBoxShadow;
  padding: 10px;
  &-title {
    font-weight: 600;
    text-align: center;
    margin-bottom: 10px;
  }
  &-canvas {
    height: 300px !important;
    width: 600px !important;
    margin: 0 auto;

    @include tablet {
      height: 200px !important;
      width: 400px !important;
    }

    @include phone {
      height: 150px !important;
      width: 300px !important;
    }
    @include small-phone {
      height: 140px !important;
      width: 280px !important;
    }
  }
}
