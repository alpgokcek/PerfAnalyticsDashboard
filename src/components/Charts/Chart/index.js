import React from "react";
import PropTypes from "prop-types";
import { Line } from "react-chartjs-2";
import EmptyContent from "./EmptyContent";

import "./chart.scss";

const Chart = (props) => {
  const { title, keyField, metrics } = props;

  const categorizedMetrics = metrics.reduce(function (r, a) {
    r[a.url] = r[a.url] || [];
    r[a.url].push(a);
    return r;
  }, Object.create(null));

  const datasets = Object.keys(categorizedMetrics).map((k) => {
    return {
      label: k,
      data: categorizedMetrics[k].map((item) => ({
        x: item.timestamp,
        y: item[keyField],
      })),
    };
  });
  const data = {
    labels: metrics.map((item) => item.timestamp).sort(),
    datasets,
  };

  return (
    <div className="chart">
      <div className="chart-title" data-test="chart-title">
        {title}
      </div>
      <div data-test={`chart-${keyField}-wrapper`}>
        {metrics.length > 0 ? (
          <Line
            className="chart-canvas"
            data={data}
            data-test={`chart-${keyField}`}
          />
        ) : (
          <EmptyContent data-test="chart-empty" />
        )}
      </div>
    </div>
  );
};

Chart.propTypes = {
  title: PropTypes.string.isRequired,
  keyField: PropTypes.string.isRequired,
  metrics: PropTypes.array.isRequired,
};

export default Chart;
