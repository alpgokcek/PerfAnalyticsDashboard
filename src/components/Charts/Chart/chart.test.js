import React from "react";
import Chart from "./index";

import Enzyme, { mount } from "enzyme";
import toJson from "enzyme-to-json";
import EnzymeAdapter from "enzyme-adapter-react-16";
import { findByTestAttribute } from "../../../test";
import { MOCK_METRIC } from "../../../test/consts";

Enzyme.configure({ adapter: new EnzymeAdapter() });

const defaultProps = {
  title: "Testing is fun!",
  keyField: "ttfb",
  metrics: [MOCK_METRIC, MOCK_METRIC, MOCK_METRIC],
};

const setup = (props) => {
  return mount(<Chart {...props} />);
};
jest.mock("react-chartjs-2", () => ({
  Line: () => <div>chart</div>,
}));
describe("snapshot and unit tests for chart", () => {
  let wrapper;
  beforeAll(() => {
    const mGetRandomValues = jest.fn().mockReturnValueOnce(new Uint32Array(10));
    Object.defineProperty(window, "crypto", {
      value: { getRandomValues: mGetRandomValues },
    });
    jest
      .useFakeTimers("modern")
      .setSystemTime(new Date("2021-05-23").getTime());
  });

  beforeEach(() => {
    wrapper = setup(defaultProps);
  });

  it("should render correctly", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render title correctly", () => {
    const inputComponent = findByTestAttribute(wrapper, "chart-title");
    expect(inputComponent.text()).toMatch(/Testing is fun!/);
  });

  it("should render chart correctly", () => {
    wrapper.setProps({ ...defaultProps });
    let inputComponent = findByTestAttribute(wrapper, "chart-ttfb-wrapper");
    expect(inputComponent.children()).toHaveLength(1);
    inputComponent = findByTestAttribute(wrapper, "chart-ttfb");
    expect(inputComponent.exists()).toBeTruthy();
    inputComponent = findByTestAttribute(wrapper, "chart-empty");
    expect(inputComponent.exists()).not.toBeTruthy();
  });

  it("should not render chart", () => {
    wrapper.setProps({ ...defaultProps, metrics: [] });
    let inputComponent = findByTestAttribute(wrapper, "chart-ttfb-wrapper");
    expect(inputComponent.children()).toHaveLength(1);
    inputComponent = findByTestAttribute(wrapper, "chart-ttfb");
    expect(inputComponent.exists()).not.toBeTruthy();
    inputComponent = findByTestAttribute(wrapper, "chart-empty");
    expect(inputComponent.exists()).toBeTruthy();
  });
});
