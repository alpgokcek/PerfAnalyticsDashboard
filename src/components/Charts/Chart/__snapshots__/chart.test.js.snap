// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot and unit tests for chart should render correctly 1`] = `
<Chart
  keyField="ttfb"
  metrics={
    Array [
      Object {
        "domLoad": "200",
        "fcp": "500",
        "timestamp": "2021-05-22T22:33:10.292Z",
        "ttfb": "700",
        "url": "https://google.com",
        "windowLoadEvents": "300",
      },
      Object {
        "domLoad": "200",
        "fcp": "500",
        "timestamp": "2021-05-22T22:33:10.292Z",
        "ttfb": "700",
        "url": "https://google.com",
        "windowLoadEvents": "300",
      },
      Object {
        "domLoad": "200",
        "fcp": "500",
        "timestamp": "2021-05-22T22:33:10.292Z",
        "ttfb": "700",
        "url": "https://google.com",
        "windowLoadEvents": "300",
      },
    ]
  }
  title="Testing is fun!"
>
  <div
    className="chart"
  >
    <div
      className="chart-title"
      data-test="chart-title"
    >
      Testing is fun!
    </div>
    <div
      data-test="chart-ttfb-wrapper"
    >
      <Line
        className="chart-canvas"
        data={
          Object {
            "datasets": Array [
              Object {
                "data": Array [
                  Object {
                    "x": "2021-05-22T22:33:10.292Z",
                    "y": "700",
                  },
                  Object {
                    "x": "2021-05-22T22:33:10.292Z",
                    "y": "700",
                  },
                  Object {
                    "x": "2021-05-22T22:33:10.292Z",
                    "y": "700",
                  },
                ],
                "label": "https://google.com",
              },
            ],
            "labels": Array [
              "2021-05-22T22:33:10.292Z",
              "2021-05-22T22:33:10.292Z",
              "2021-05-22T22:33:10.292Z",
            ],
          }
        }
        data-test="chart-ttfb"
      >
        <div>
          chart
        </div>
      </Line>
    </div>
  </div>
</Chart>
`;
